File: helpers/wikimedia.py
from typing import Any

import requests

from helpers.dynamodb import DynamoDBWrapper
from helpers.http import wikimedia_headers

dynamodb = DynamoDBWrapper()


def _make_request(req):
    result = requests.get(
        "https://commons.wikimedia.org/w/api.php",
        params=req,
        headers=wikimedia_headers(),
    ).json()
    if "error" in result:
        raise Exception(result["error"])
    if "warnings" in result:
        print(result["warnings"])
    if "query" in result:
        return result
    else:
        raise Exception("Something went wrong!")


def _find_non_posted_image(results) -> None | Any:
    """
    Returns nothing if all the results have already been posted.
    Otherwise, returns the id and title of an image that has not been posted.
    """
    for result in results:
        if not dynamodb.is_already_posted(result["pageid"]):
            return result
    return None


##
# Returns a title and an ID, which can be used to query for the image itself.
def get_random_image() -> dict[str, int]:
    request = {
        "action": "query",
        "format": "json",
        "list": "search",
        "srsearch": 'incategory:"Quality images"',
        "srnamespace": "6",  # File namespace
        "srsort": "random",
        "srlimit": "max",
    }
    while True:
        result = _make_request(request)
        results = result["query"]["search"]

        for result in results:
            if not dynamodb.is_already_posted(result["pageid"]):
                return {
                    "pageid": result["pageid"],
                    "title": result["title"],
                }


def get_file_details(file_title: str) -> dict[str, any]:
    request = {
        "action": "query",
        "format": "json",
        "titles": [file_title],
        "prop": "imageinfo",
        "iiprop": "extmetadata|url",
    }
    result = requests.get(
        "https://commons.wikimedia.org/w/api.php",
        params=request,
        headers=wikimedia_headers(),
    ).json()
    result_list = list(result["query"]["pages"].values())
    print(result_list)
    return result_list[0]["imageinfo"][0]

File: src/helpers/wikimedia.ts
import axios from 'axios';
import { addWeeks, subSeconds } from 'date-fns';

const WIKIMEDIA_API = 'https://commons.wikimedia.org/w/api.php';

interface WikimediaFile {
  pageid: number;
  title: string;
}

interface WikimediaFileDetails {
  url: string;
  descriptionurl: string;
}

const randomTime = (
  start: Date = new Date(2011, 2, 8, 13),
  end: Date = subSeconds(new Date(), addWeeks(new Date(), 4).getTime())
): Date => {
  const delta = end.getTime() - start.getTime();
  const randomSecond = Math.floor(Math.random() * delta);
  return new Date(start.getTime() + randomSecond);
};

const makeRequest = async (params: Record<string, any>) => {
  const { data } = await axios.get(WIKIMEDIA_API, { params });
  if ('error' in data) {
    throw new Error(JSON.stringify(data.error));
  }
  if ('warnings' in data) {
    console.warn(data.warnings);
  }
  if ('query' in data) {
    return data;
  }
  throw new Error('something went wrong!');
};

const findNonPostedImage = (results: WikimediaFile[]): WikimediaFile => {
  return results[0]; // in the future we could check if it's been posted
};

export const getRandomImage = async (): Promise<WikimediaFile> => {
  const request = {
    action: 'query',
    format: 'json',
    list: 'categorymembers',
    cmtype: 'file',
    cmtitle: 'Category:Quality_images',
    cmstart: randomTime(),
    cmsort: 'timestamp',
    cmdir: 'ascending',
  };

  let lastContinue = {};
  while (true) {
    const req = { ...request, ...lastContinue };
    const result = await makeRequest(req);

    if ('query' in result) {
      const results = result.query.categorymembers;
      const freshResult = findNonPostedImage(results);
      if (freshResult) {
        return freshResult;
      }
    }
    lastContinue = result.continue;
  }
};

export const getFileDetails = async (
  fileTitle: string
): Promise<WikimediaFileDetails> => {
  const request = {
    action: 'query',
    format: 'json',
    titles: fileTitle,
    prop: 'imageinfo',
    iiprop: 'url',
  };

  const { data } = await axios.get(WIKIMEDIA_API, { params: request });
  const resultList = Object.values(data.query.pages);
  console.log(resultList);
  return (resultList[0] as any).imageinfo[0];
};
